runtime/compiler/runtime/JITClientSession.cpp

```diff
@@ -308,11 +330,54 @@ ClientSessionData::getOrCacheVMInfo(JITServer::ServerStream *stream)
    if (!_vmInfo)
       {
       stream->write(JITServer::MessageType::VM_getVMInfo, JITServer::Void());
-      _vmInfo = new (PERSISTENT_NEW) VMInfo(std::get<0>(stream->read<VMInfo>()));
+      auto recv = stream->read<VMInfo, std::vector<uintptr_t>, std::vector<uintptr_t> >();
+      _vmInfo = new (PERSISTENT_NEW) VMInfo(std::get<0>(recv));
+      _vmInfo->_j9SharedClassCacheDescriptorList = reconstructJ9SharedClassCacheDescriptorList(std::get<1>(recv), std::get<2>(recv));
       }
    return _vmInfo;
    }
 
+J9SharedClassCacheDescriptor *
+ClientSessionData::reconstructJ9SharedClassCacheDescriptorList(std::vector<uintptr_t>& listOfCacheStartAddress, std::vector<uintptr_t>& listOfCacheSizeBytes)
+   {
+   J9SharedClassCacheDescriptor * prev = NULL;
+   J9SharedClassCacheDescriptor * first = NULL;
+   J9SharedClassCacheDescriptor * cur = NULL;
+   for (int i = 0; i < listOfCacheStartAddress.size(); i++)
+      {
+      cur = new (PERSISTENT_NEW) J9SharedClassCacheDescriptor();
+      cur->cacheStartAddress = (J9SharedCacheHeader*)(listOfCacheStartAddress[i]);
+      cur->cacheSizeBytes = listOfCacheSizeBytes[i];
+      if (prev)
+         prev->previous = cur;
+      else
+         first = cur;
+      prev = cur;
+      }
+   // Mark the previous pointer of last node to be the first node
+   // so last->previous == first
+   if (prev)
+      prev->previous = first;
+   return prev;
+   }
```

Comment:
You should set both the `previous` and `next` links here so that the list is well-formed and you need to guard the part that sets `previous` with an `#ifdef`. Someone could easily add code to this class that assumes `next` is always valid and introduce a bug.

Reply:
Forget about the `#ifdef` stuff, it looks like https://github.com/eclipse/openj9/commit/031af05cd8f87972e9425597a0f781c621597445 was revised and no longer has the preprocessor guard around the `previous` element so it will always exist.